### PR TITLE
Improve i18n

### PR DESCRIPTION
**Pull request by pamasse**
Like PR #116, this PR try to improve I18n support by using some new tool to align the translation files ({en,fr,da,sv}.json) to the code.
In order to use ngx-extractor, we need to use correctly our translator tool. So, I update some piece of codes to correctly use the translator service (this.translateService.get()).

I add two commands to the project:
-  i18n:init : Create a template.json file with all keys used in the code 
- i18n:extract: Update {en,fr,da,sv}.json files with news keys if detected or delete used ones.

to use these command, simply run `npm run i18n:extract` for example.

### CODE REVIEW

**Comment by matsduf** [comment]
@pamasse, you write "with news keys if detected or delete used ones". Should it be "with news keys if detected or delete **_unused_** ones"?

Please tag with label.

Is this fixing some bugs that we should include in v2019.2, or is this an improvement that could wait for v2020.1? Please set milestone.

There are a lot of changes in the "i18n/**.json" files. Are there any changes of contents, or just format?

**Review by sandoche2k — DISMISSED** [review]

**Comment by sandoche2k** [comment]
@pnax do you think you can review this issue?

**Comment by pamasse** [comment]
Hi @matsduf,

> @pamasse, you write "with news keys if detected or delete used ones". Should it be "with news keys if detected or delete **_unused_** ones"?

Yes, the script deleted keys that are defined in the json file and not used in the application source code. 

> Please tag with label.
> Is this fixing some bugs that we should include in v2019.2, or is this an improvement that could wait for v2020.1? Please set milestone.
Noted for the next PR ! 


> There are a lot of changes in the "i18n/**.json" files. Are there any changes of contents, or just format?

It's just format modifications. The files are now generated from scripts.

**Review by matsduf — DISMISSED** [review]
This is quite large, but I do not see any issue. Testing will have to verify all the details.

After merge, fr.json will look like
```
{
	"About DNS Text": "La plupart des gens qui utilisent Internet n'ont jamais réfléchi à ce qui se passe réellement quand on tape une adresse dans leur navigateur Web, puis clique sur le bouton Entrée. Pour simplifier, le système de noms de domaine (DNS) est une base de données qui associe des noms de domaine à des adresses IP, de la même manière qu'un annuaire téléphonique associe des noms à des numéros de téléphone. Une adresse IP est une série unique de chiffres qui identifie chaque ordinateur connecté à Internet. C'est semblable à la façon dont chaque téléphone a son propre numéro dans le réseau téléphonique. Une longue série de chiffres fonctionne bien pour les ordinateurs, mais il est difficile pour les gens de mémoriser. Par conséquent, DNS a été développé afin de permettre d'atteindre le serveur web d'un domaine spécifique ou un serveur de messagerie auquel nous voulons envoyer un email.",
	"About ZoneMaster": "À propos de Zonemaster",
	"About ZoneMaster Text": "Donnez à votre domaine un check up complet! Zonemaster vous aide à contrôler le fonctionnement de votre site Web. En prime, vous obtenez également une meilleure compréhension du fonctionnement du système de noms de domaine, le DNS. Lorsqu'un domaine (également appelé zone) est envoyé à Zonemaster, le programme examine l'état du domaine du début à la fin. Cela se fait via Zonemaster en examinant DNS de la racine (.) vers le TLD (domaine de premier niveau, par exemple, .fr), puis enfin via les serveurs DNS qui contiennent des informations sur le domaine spécifié (par exemple, zonemaster.fr ). Zonemaster est également capable d'effectuer de nombreux autres tests, par exemple, vérifier les signatures DNSSEC, que différents hôtes peuvent être accédés et que les adresses IP sont valides. Tout est fait pour s'assurer que votre domaine fonctionne aussi bien que possible.",
	"About domain name system, DNS": "À propos du systeme de nom de domaine, DNS (Domain Name System)",
	"Algorithm": "Algorithme",
```
which is quite hard to read and manage. Would it work to have a more structured format as below? Would that be better?

A second question is if we should have some automatic tool that verifies that all xx.json have the same keys?

```
{
	"About DNS Text": 
		"La plupart des gens qui utilisent Internet n'ont jamais réfléchi à ce qui se passe
		réellement quand on tape une adresse dans leur navigateur Web, puis clique sur 
		le bouton Entrée. Pour simplifier, le système de noms de domaine (DNS) est une 
		base de données qui associe des noms de domaine à des adresses IP, de la même
		manière qu'un annuaire téléphonique associe des noms à des numéros de téléphone. 
		Une adresse IP est une série unique de chiffres qui identifie chaque ordinateur 
		connecté à Internet. C'est semblable à la façon dont chaque téléphone a son propre
		numéro dans le réseau téléphonique. Une longue série de chiffres fonctionne bien 
		pour les ordinateurs, mais il est difficile pour les gens de mémoriser. Par conséquent, 
		DNS a été développé afin de permettre d'atteindre le serveur web d'un domaine
		spécifique ou un serveur de messagerie auquel nous voulons envoyer un email.",
	"About ZoneMaster": 
		"À propos de Zonemaster",
	"About ZoneMaster Text": 
		"Donnez à votre domaine un check up complet! Zonemaster vous aide à contrôler 
		le fonctionnement de votre site Web. En prime, vous obtenez également une meilleure 
		compréhension du fonctionnement du système de noms de domaine, le DNS. Lorsqu'un 
		domaine (également appelé zone) est envoyé à Zonemaster, le programme examine 
		l'état du domaine du début à la fin. Cela se fait via Zonemaster en examinant DNS de 
		la racine (.) vers le TLD (domaine de premier niveau, par exemple, .fr), puis enfin via 
		les serveurs DNS qui contiennent des informations sur le domaine spécifié (par 
		exemple, zonemaster.fr ). Zonemaster est également capable d'effectuer de 
		nombreux autres tests, par exemple, vérifier les signatures DNSSEC, que différents 
		hôtes peuvent être accédés et que les adresses IP sont valides. Tout est fait pour 
		s'assurer que votre domaine fonctionne aussi bien que possible.",
	"About domain name system, DNS": 
		"À propos du systeme de nom de domaine, DNS (Domain Name System)",
	"Algorithm": 
		"Algorithme",
```

**Comment by matsduf** [comment]
> > There are a lot of changes in the "i18n/**.json" files. Are there any changes of contents, or just format?
> 
> It's just format modifications. The files are now generated from scripts.

Generated from what? Do you say that the primary data is somewhere else? I do not find it.

**Comment by matsduf** [comment]
This fails in Travis, but #122 does not. Does that mean that the issue has been resolved by #122?

File: src/app/components/result/result.component.ts [hunk 1]
```diff
@@ -92,21 +92,12 @@ export class ResultComponent implements OnInit, OnChanges {
...
-  private getDismissReason(reason: any): string {
-    if (reason === ModalDismissReasons.ESC) {
-      return 'by pressing ESC';
-    } else if (reason === ModalDismissReasons.BACKDROP_CLICK) {
-      return 'by clicking on a backdrop';
-    } else {
-      return  `with: ${reason}`;
-    }
-  }
+$
```
Comment (pnax) [inline src/app/components/result/result.component.ts:100]: Just wondering if this dollar sign has any meaning or if it lies here by mistake

File: src/app/services/dns-check.service.ts [hunk 2]
```diff
@@ -2,19 +2,22 @@ import { Injectable } from '@angular/core';
...
-  constructor(private alertService: AlertService, private http: HttpClient) {
+  constructor(private alertService: AlertService, 
+    private http: HttpClient,
+    private translateService: TranslateService) {
     this.backendUrl = AppService.apiEndpoint();
     this.clientInfo = AppService.getClientInfo();
 
     if (this.backendUrl) {
-      console.error('Please set the api endpoint');
+      console.error(this.translateService.get('Please set the api endpoint'));
```
Comment (pnax) [inline src/app/services/dns-check.service.ts:20]: Why using a different syntax here ? And not, as in the other files, something like :
```
this.translateService.get('Please set the api endpoint').subscribe((res: string) => {
  console.error(res);
});
```

**Review by pnax — COMMENTED** [review]
> @pnax do you think you can review this issue?

yep, just done it :)

> Generated from what? Do you say that the primary data is somewhere else? I do not find it.

From what I understood of how [NGX-Translate](http://www.ngx-translate.com/) works, I think that the files are generated from the code itself. Hence the new lines in the `package.json` file I think.

**Comment by pamasse** [comment]
hi,

it's exact, the keys in the translation files are generated from the
code itself.

The dollar sign seems to be a mistake.

The last point is a good update, could you do a PR for that ?


Le jeu. 24 sept. 2020 à 17:57, pnax <notifications@github.com> a écrit :
>
> @pnax commented on this pull request.
>
> @pnax do you think you can review this issue?
>
> yep, just done it :)
>
> Generated from what? Do you say that the primary data is somewhere else? I do not find it.
>
> From what I understood of how NGX-Translate works, I think that the files are generated from the code itself. Hence the new lines in the package.json file I think.
>
> ________________________________
>
> In src/app/components/result/result.component.ts:
>
> >      });
>    }
> -
> -  private getDismissReason(reason: any): string {
> -    if (reason === ModalDismissReasons.ESC) {
> -      return 'by pressing ESC';
> -    } else if (reason === ModalDismissReasons.BACKDROP_CLICK) {
> -      return 'by clicking on a backdrop';
> -    } else {
> -      return  `with: ${reason}`;
> -    }
> -  }
> +$
>
> Just wondering if this dollar sign has any meaning or if it lies here by mistake
>
> ________________________________
>
> In src/app/services/dns-check.service.ts:
>
> >      this.backendUrl = AppService.apiEndpoint();
>      this.clientInfo = AppService.getClientInfo();
>
>      if (this.backendUrl) {
> -      console.error('Please set the api endpoint');
> +      console.error(this.translateService.get('Please set the api endpoint'));
>
> Why using a different syntax here ? And not, as in the other files, something like :
>
> this.translateService.get('Please set the api endpoint').subscribe((res: string) => {
>   console.error(res);
> });
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.



-- 
Pierre-Aymeric MASSE

**Comment by matsduf** [comment]
> hi, it's exact, the keys in the translation files are generated from the code itself.

@pamasse,

Can you please explan. I guees that you mean that the keys in the xx.json files come from the code, but the data (translation into French, Swedish etc), where does that come from?

In a comment I asked for a different formatting of the xx.json files because that I assumed that they should be updated when translations are added or corrected. Is that correct?

Can you also please comment my suggestion of a different formatting?

**Comment by pamasse** [comment]
Hi @matsduf !

> Can you please explain. I guess that you mean that the keys in the xx.json files come from the code, but the data (translation into French, Swedish, etc), where does that come from?
> In a comment I asked for different formatting of the xx.json files because that I assumed that they should be updated when translations are added or corrected. Is that correct?
I proposed to use [NGX-Translate-extract](https://github.com/biesbjerg/ngx-translate-extract) which try to extract the translation keys from the code itself (example: `{{'About ZoneMaster'|translate}}`, the key is About ZoneMaster) and check if the key is in all translation files en|fr|sv|da.json (it check all JSON files in src/assets/i18n folder).
If the key is not in the file, ngx-translate-extract add it in the file, if a key is already in the file, nothing is changed. 
Then, it checks if the keys in the files are really used in the code. If a key is not used, it deletes the corresponding line in all translation files.
Finally, it sorts the keys in the translations files by alphabetical order. 

> Can you also please comment my suggestion of a different formatting?
Some bad news, is not possible to do what you want to do. Json doesn't allow this kind of format for JSON, multiline is not allowed. 
The best that I can propose to you is to create a custom script to change the current format to something like :
```
{
	"About DNS Text": 
		"La plupart des gens qui utilisent Internet n'ont jamais réfléchi à ce qui se passe \
réellement quand on tape une adresse dans leur navigateur Web, puis clique sur \
le bouton Entrée. Pour simplifier, le système de noms de domaine (DNS) est une \
base de données qui associe des noms de domaine à des adresses IP, de la même \
manière qu'un annuaire téléphonique associe des noms à des numéros de téléphone. \ 
Une adresse IP est une série unique de chiffres qui identifie chaque ordinateur \
connecté à Internet. C'est semblable à la façon dont chaque téléphone a son proper \
numéro dans le réseau téléphonique. Une longue série de chiffres fonctionne bien \
pour les ordinateurs, mais il est difficile pour les gens de mémoriser. Par conséquent, \
DNS a été développé afin de permettre d'atteindre le serveur web d'un domaine \
spécifique ou un serveur de messagerie auquel nous voulons envoyer un email.",
	"About ZoneMaster": 
		"À propos de Zonemaster"
}
```

Do you want this new format?

**Comment by pnax** [comment]
Hello @pamasse 

> The last point is a good update, could you do a PR for that ?

see https://github.com/pamasse/zonemaster-gui/pull/1

**Comment by matsduf** [comment]
> I proposed to use [NGX-Translate-extract](https://github.com/biesbjerg/ngx-translate-extract) which try to extract the translation keys from the code itself (example: `{{'About ZoneMaster'|translate}}`, the key is About ZoneMaster) and check if the key is in all translation files en|fr|sv|da.json (it check all JSON files in src/assets/i18n folder).
> If the key is not in the file, ngx-translate-extract add it in the file, if a key is already in the file, nothing is changed.
> Then, it checks if the keys in the files are really used in the code. If a key is not used, it deletes the corresponding line in all translation files.
> Finally, it sorts the keys in the translations files by alphabetical order.

That sounds fine. Could you document the process including pointing at site where you can download the program? I think it should be in [ReleaseProcess-preparation](https://github.com/zonemaster/zonemaster/blob/develop/docs/internal-documentation/maintenance/ReleaseProcess-preparation.md) or in Zonemaster-GUI and linked from the document.

In most cases the key is the English text to be displayed, and that is straight forward. When you translated you start from the key.

In some cases the key is a key for a longer text, e.g. "About DNS Text", and there is a longer text in all xx.json files, also in the en.json. I think the key, then, should be more key like, e.g. "ABOUT-DNS-TEXT" to make it more obvious. There should be no problem with these cases.

The third category are cases where the key could be a message, but is "translated" to a different messages also in Enlish, e.g. "About ZoneMaster" --> "About Zonemaster". We have had that situation in Engine for the messages too. We should really get rid of those. In en.json there should only be the two first categories, key == value or key == SPECIAL-KEY.

> > Can you also please comment my suggestion of a different formatting?
> 
> Some bad news, it is not possible to do what you want to do. JSON doesn't allow this kind of format for JSON, multiline is not allowed.

I understand. We have to accept the format. Maybe there is some tool that could help the translator. PO files are more complex, but there are tools that make the work easier.

I see that the tool gives some options on the output and maybe some other combination might give a more readable result.
```
Output
  --format, -f                Format    [string] [choices: "json", "namespaced-json", "pot"] [default: "json"]
  --format-indentation, --fi  Format indentation (JSON/Namedspaced JSON)                [string] [default: "	"]
```

**Review by matsduf — APPROVED** [review]
Documentation and keys are handled in a separate issue.

**Comment by matsduf** [comment]
A follow-up issue has been created: #143.

**Pull request merged**